### ISSUE TITLE
Program crashes

### ISSUE THREAD

**Issue by rwhitworth** [ericmandel/funtools#13]
Using American Fuzzy Lop (afl-fuzz) to compile and execute various funtools programs has led to quite a few crashing conditions.  Is fixing these something you are interested in?  Examples of input that causes crashes can be found here: https://github.com/rwhitworth/funtools-fuzz.  The README in each directory also contains the command line used, in case that affects the crash.

Thanks.

**Comment by ericmandel** [comment]
@rwhitworth Fascinating! I don't actively develop funtools anymore, but lots of people are still using it and I am willing to put some time into fixing crashes.

Can you give me some hints about how to use your results? For example, does this mean anything by itself?
````
 funtools-fuzz/funcen/id:000000,sig:11,src:000057,op:havoc,rep:2

 3 lines (2 sloc) 14 Bytes
::::.	q
::.qq
````
How do I start? I built AFL on Linux -- am I now supposed to run it myself to get more detailed output?

**Comment by rwhitworth** [comment]
@ericmandel Thanks for the fast response and willingness to look at these input files.  You don't need AFL to utilize these input files as-is.  AFL is only needed if you would like to continue fuzz testing in the future (verify correctness of tools not fuzzed, verify fixes, etc)

I just noticed the tools don't like input files with odd characters in their filenames and that may be causing problems when you tried to reproduce?

An example of a crash is to run this from the top level of the funtools-fuzz repo:
```
cp funcen/id\:000000\,sig\:11\,src\:000057\,op\:havoc\,rep\:2 funcen/id000000
funcen funcen/id000000 'cir(512,512,10);cir 450 450 10'
Segmentation fault
```

Using `dmesg | grep funcen` produces this as the segfault details:
`[86119.099335] funcen[29534]: segfault at 0 ip 00007fcb3448cfc3 sp 00007ffd389d5048 error 4 in libc-2.19.so[7fcb3440d000+1a1000]`

(btw, I'm all ears if there is a better way to gather details about a segfault or the cause of a segfault.  I'm sure gdb is the answer but not familiar enough with it)

The funcen binary used in the above example was created from the github source and compiled on Debian Jessie with gcc (nothing special and no AFL compiler used)

**Comment by ericmandel** [comment]
@rwhitworth Ah, I see: I didn't understand that you actually were supplying the input files ... that makes it easy ...

I added some validity checks in the text file open code, and ran both funcen and fundisp on all of your tests files without any crashes. Everything is updated in GitHub, so please feel free to report other problems. This really has been interesting and I learned something about the limits of my view of testing.

May I ask why you are testing the now-unsupported funtools package? As part of a Debian test suite?

**Comment by rwhitworth** [comment]
Those changes did the trick.  I ran AFL again and no crash causing inputs were identified.

I learned of the project from the Debian project website.  I don't have an interesting in astronomy, bur rather in software quality, which is my day job.  The name of the project caught my eye when looking for something to learn about (code structure, how to test something I'm completely unfamiliar with, etc).  The resulting commits to fix the crashes was more succinct than I had anticipated and shown I have much more to learn about software development.

Thank you for your work on this project and your open source contributions.
